src/app/3d/qgs3dmaptoolmeasureline.cpp

```diff
@@ -231,3 +177,60 @@ QVector<QgsPoint> Qgs3DMapToolMeasureLine::points() const
 {
   return mPoints;
 }
+
+void Qgs3DMapToolMeasureLine::mousePressEvent( QMouseEvent *event )
+{
+  mMouseHasMoved = false;
+  mMouseClickPos = event->pos();
+}
+
+void Qgs3DMapToolMeasureLine::mouseMoveEvent( QMouseEvent *event )
+{
+  if ( !mMouseHasMoved &&
+       ( event->pos() - mMouseClickPos ).manhattanLength() >= QApplication::startDragDistance() )
+  {
+    mMouseHasMoved = true;
+  }
+
+  if ( mPoints.isEmpty() || mDone )
+    return;
+
+  const QgsRay3D ray = Qgs3DUtils::rayFromScreenPoint( event->pos(), mCanvas->windowSize(), mCanvas->cameraController()->camera() );
+  const float dist = static_cast<float>( mPoints.last().z() - ray.origin().y() ) / ray.direction().y();
+  const QVector3D hoverPoint = ray.origin() + ray.direction() * dist;
+  const QgsVector3D mapCoords = Qgs3DUtils::worldToMapCoordinates( hoverPoint, mCanvas->map()->origin() );
+  mRubberBand->moveLastPoint( QgsPoint( mapCoords.x(), mapCoords.y(), mapCoords.z() ) );
```

Comment:
the coordinates of the last point (the one that moves) are not calculated correctly if the terrain has vertical scale applied

Reply:
fixed